pages/blog/the-anatomy-of-a-graphql-request.mdx

```diff
@@ -0,0 +1,341 @@
+  image: '/blog-assets/the-anatomy-of-a-graphql-request/image.png'
+}
+
+export default withArticle({ ...meta })
+
+On a high level GraphQL servers are quite easy to use. You pass it a GraphQL schema, start the server and that is it.
+For many use-cases that is sufficient enough.
+
+But let's dive deeper, under the hood and learn what is exactly happening under the hood and why it is useful to hook into each of the phases of an GraphQL request!
+
+While GraphQL can be done over almost any protocol, this article focuses on the most commonly used protocol [GraphQL over HTTP](https://github.com/graphql/graphql-over-http). However, most knowledge can be transfered to other protocols such as GraphQL over WebSockets or other more exotic ones.
+
+![The Phases of an GraphQL Request](/blog-assets/the-anatomy-of-a-graphql-request/phases-breakdown.png)
+
+## HTTP Parsing and Normalization
+
+Clients send HTTP requests to the server with a payload that contains an operation document string (query, mutation or subscription) and optionally some variables and the operation name from the document that shall be executed.
+
+Generally it is a JSON object for most requests that are executed via the POST http method.
+
+**Example JSON POST Body**
+
+```json
+{
+  "query": "query UserById($id: ID!) { user(id: $id) { id name } }",
+  "variables": { "id": "10" },
+  "operationName": "UserById"
+}
+```
+
+However, when using the GET HTTP method (e.g. for query operations) those parameters can also be provided as a query search string. The values are then URL-encoded.
+
+**Example GET URL**
+
+```
+http://localhost:8080/graphql?query=query%20UserById%28%24id%3A%20ID%21%29%20%7B%20user%28id%3A%20%24id%29%20%7B%20id%20name%20%7D%20%7D&variables=%7B%20%22id%22%3A%20%2210%22%20%7DoperationName=UserById
+```
+
+The GraphQL HTTP servers first task is to parse and normalize the body or query string and furthermore determine the protocol that shall be used for sending the response. Currently the protocols can be `application/graphql+json` (or `application/json` for legacy clients) for a single result that yields from the execution phase, `multipart/mixed` for incremental delivery (when using `@defer` and `@stream` directives). Not officially in the specification, but also used is `text/event-stream` for event streams (e.g. when executing subscription or live query operations).
+
+## GraphQL parse
+
+After parsing and normalizing the request the server will pass on the GraphQL parameters onto the GraphQL engine which will first parse the GraphQL operation document (which can contain any number of query, mutation or subscription operations and fragment definitions).
+
+If there is any typo or syntax error this phase will yield GraphQLErrors for each of those issues and pass them back to the server layer, so it can send those back to the client.
+
+For the following invalid GraphQL operation (`)` missing after `$id`):
+
+```
```

Comment:
```gql

Reply:
I omitted it here because it is an invalid operation